core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -0,0 +1,561 @@
+ * -----------------------------------------------------------------------------|
+ * 2 | Hash  - FULL entry hash index, which might be different (bigger) than    |
+ *   |         entry hash index                                                 |
+ * ---------------------------------------------------------------------------------------
+ * 3 | Key Reference          | Deleted reference (still) encoding all info     |
+ *   |                        | needed to access the key off-heap               |
+ * -----------------------------------------------------------------------------| Deleted entry with
+ * 4 | Value Reference        | Deleted reference(still)  encoding all info     | entry hash index
+ *   |                        | needed to access the key off-heap.              | 1
+ * -----------------------------------------------------------------------------|
+ * 5 | Hash  - FULL entry hash index, which might be different (bigger) than    |
+ *   |         entry hash index                                                 |
+ * ---------------------------------------------------------------------------------------
+ * 6 | Key Reference          | Invalid key reference                           | empty entry
+ * -----------------------------------------------------------------------------|
+ * 7 | Value Reference        | Invalid value reference                         |
+ * -----------------------------------------------------------------------------|
+ * 8 | 000...00                                                                 |
+ * ---------------------------------------------------------------------------------------
+ * ...
+ *
+ *
+ * Internal class, package visibility
+ */
+class EntryHashSet<K, V> extends EntryArray<K, V> {
+
+    /*-------------- Constants --------------*/
+    static final int INVALID_FULL_HASH = 0; // because memory is initially zeroed
+    static final int DEFAULT_COLLISION_CHAIN_LENGTH = 3;
+    // HASH - the full hash index of this entry (one integer, all bits).
+    private static final int HASH_FIELD_OFFSET = 2;
+    private static final int FULL_HASH_BITS = 32; // Hash needs to be an integer
+    // One additional to the key and value references field, which keeps the full hash index of the entry
+    private static final int ADDITIONAL_FIELDS = 1;  // # of primitive fields in each item of entries array
+
+    // number of entries candidates to try in case of collision
+    private AtomicInteger collisionChainLength = new AtomicInteger(DEFAULT_COLLISION_CHAIN_LENGTH);
+
+    private final OakComparator<K> comparator;
+
+    // use reference codec to encode Full Hash Number integer (first) with its updates counter (second)
+    private final ReferenceCodec hashCodec = new ReferenceCodec(
+        FULL_HASH_BITS, // bits# to represent full hash number as integer, bits# to represent update
+        ReferenceCodec.INVALID_BIT_SIZE); // counter are calculated upon previous parameters (also int)
```

Comment:
I like that you found re-use of this code for other purposes. 
In this case, it no longer serves as a reference codec.
In actually more like a [union type](https://en.wikipedia.org/wiki/Union_type).
Maybe we should extract the "union" part from `ReferenceCodec` to a separate `Union` class and use it here.
Then `ReferenceCodec` can extend `Union`.

Reply:
done